<table align="left" width=100%>
    <tr>
        <td width="20%">
            <img src="GL-2.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Faculty Notebook <br> ( Day 3) </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## Table of Content

1. **[Import Libraries](#lib)**
2. **[One-way & Two Way ANOVA](#1way)**
    - 2.1 - **[Post-hoc Analysis](#post-hoc)**
    - 2.2 - **[Non Parametric Test (Additional Content)](#non-para)**

<a id="lib"></a>
# 1. Import Libraries

**Let us import the required libraries.**

In [8]:
# import 'pandas' 
import pandas as pd 

# import 'numpy' 
import numpy as np

# import subpackage of matplotlib
import matplotlib.pyplot as plt

# import 'seaborn'
import seaborn as sns

# to suppress warnings 
from warnings import filterwarnings
filterwarnings('ignore')

# import 'stats' package from scipy library
from scipy import stats

# import statistics to perform statistical computations
import statistics

# to test the normality 
from scipy.stats import shapiro

# import statsmodels
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

# import the functions to perform Chi-square tests
from scipy.stats import chi2_contingency
from scipy.stats import chi2
from scipy.stats import chisquare

# function to perform post-hoc test
import statsmodels.stats.multicomp as mc

In [2]:
# set the plot size using 'rcParams'
# once the plot size is set using 'rcParams', it sets the size of all the forthcoming plots in the file
# pass width and height in inches to 'figure.figsize' 
plt.rcParams['figure.figsize'] = [15,8]

In [3]:
# F Critical...

stats.f.ppf(0.95, dfn = 2, dfd = 6)

5.143252849784718

<a id="1way"></a>
# 2. One-way ANOVA

It is used to check the equality of population means for more than two independent samples. Each group is considered as a `treatment`. It assumes that the samples are taken from normally distributed populations. To check this assumption we can use the `Shapiro-Wilk Test.` Also, the population variances should be equal; this can be tested using the `Levene's Test`.

The null and alternative hypothesis is given as:
<p style='text-indent:20em'> <strong> $H_{0}$: The averages of all treatments are the same. </strong></p>
<p style='text-indent:20em'> <strong> $H_{1}$: At least one treatment has a different average. </strong></p>

Consider there are `t` treatments and `N` number of total observations. The test statistic is given as:
<p style='text-indent:28em'> <strong> $F = \frac{MTrSS}{MESS} $</strong></p>

Where,<br>
MTrSS = $\frac{TrSS}{df_{Tr}}$<br>

TrSS = $\sum_{i}^{t}\sum_{j}^{n_{i}}n_{i}(\bar{x_{i}}. - \bar{x}..)$<br> $n_{i}$ is the number of observations in $i^{th}$ treatment. <br>$\bar{x_{i}}.$ is the mean over $i^{th}$ treatment <br> $\bar{x}..$ is the grand mean (i.e. mean of all the observations). <br>

$df_{Tr}$ is the degrees of freedom for treatments (= $t-1$)

MESS = $\frac{ESS}{df_{e}}$<br>

ESS = $\sum_{i}^{t}\sum_{j}^{n_{i}}{(x_{ij} - \bar{x_{i}}.)}^{2}$

$df_{e}$ is the degrees of freedom for error (= $N-t$)

Under $H_{0}$, the test statistic follows F-distribution with ($t-1,  N-t$) degrees of freedom.


### Difference between One Way anova and Two Way Anova

* One Way Anova is when you introduce One Categorical Variable & A Numerical Variable.
* On the other hand, if we want to analyse multiple categorical variables with a numerical variable, then it is called Two Way Anova.

* For example Pclass vs Fare is an example of One Way Anova with One categorical variable with 03 Levels (Class1, Class2 & 3) and a Numerical Variable.

* Whereas the Two Way Anova is two or more categorical variables with a numerical variables.

Let us calculate the F values for different levels of significance ($\alpha$).

In [4]:
# let us find the F-values for different alpha values with (10,10) degrees of freedom

# create an empty dataframe to store the alpha and corresponding F-value
df_F = pd.DataFrame()

# create a dictionary of different alpha values
alpha =  [0.25, 0.1, 0.05, 0.01] 

# use for loop to calculate the F-value for each alpha value
for i in range(len(alpha)):
    
    # use 'stats.f.isf()' to find the F-value corresponding to the upper tail probability 'q'
    # pass the value of 'alpha' to the parameter 'q'
    # pass the (10,10) degrees of freedom to the parameter 'dfn' and 'dfd' respectively
    # use 'round()' to round-off the value to 2 digits
    f_val = np.abs(round(stats.f.isf(q = alpha[i], dfn = 10, dfd = 10), 2))

    # create a dataframe using dictionary to store the alpha and corresponding F-value
    # set the loop iterator 'i' as the index of the dataframe
    row =  pd.DataFrame({"alpha": alpha[i], "F": f_val}, index = [i])

    # append the row to the dataframe 'df_F'
    df_F = df_F.append(row)

# print the final dataframe
df_F

alpha     F
0   0.25  1.55
1   0.10  2.32
2   0.05  2.98
3   0.01  4.85

### Lets Solve for Anova. Here is a Sample Dataset with 3 Categories. Lets apply Anova step by step

In [8]:
cat1 = [1,2,5]
cat2 = [2,4,2]
cat3 = [2,3,4]

In [9]:
# F Critical
stats.f.isf(0.05, dfn = 2, dfd = 6)

5.143252849784718

In [10]:
# F Test
stats.f_oneway(cat1, cat2, cat3)

F_onewayResult(statistic=0.049999999999999996, pvalue=0.9516215013591449)

### Example:

#### 1. Total marks in aptitude exam are recorded for students with different race/ethnicity. Test whether all the races/ethnicities have an equal average score with 0.05 level of significance. 

Use the performance dataset of students available in the CSV file `students_data.csv`.

In [27]:
# read the students performance data 
df_student = pd.read_csv('StudentsPerformance (2).csv')

# display the first two observations
df_student.head(2)

gender race/ethnicity     lunch test preparation course  math score  \
0  female        group B  standard                    none          89   
1  female        group C  standard               completed          55   

   reading score  writing score  total score training institute  
0             55             56          200    Nature Learning  
1             63             72          190    Nature Learning

The null and alternative hypothesis is:

H<sub>0</sub>: The average score of all races/ethnicities is same<br>
H<sub>1</sub>: At least one race/ethnicity has a different average score

In [28]:
# Find the Levels in Race
df_student["race/ethnicity"].unique()

array(['group B', 'group C', 'group A', 'group D', 'group E'],
      dtype=object)

In [29]:
import statsmodels.formula.api as sfa

There are total 5 unique race/ethnicity in the dataset.

In [30]:
# Statistical Library
df_student["total_score"] = df_student["total score"]
df_student["race"] = df_student["race/ethnicity"]

In [31]:
# Regression
model = sfa.ols("total_score~race", data = df_student).fit()

In [32]:
# Regression Output
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            total_score   R-squared:                       0.003
Model:                            OLS   Adj. R-squared:                 -0.001
Method:                 Least Squares   F-statistic:                    0.7891
Date:                Fri, 27 Oct 2023   Prob (F-statistic):              0.532
Time:                        12:22:35   Log-Likelihood:                -4560.8
No. Observations:                1000   AIC:                             9132.
Df Residuals:                     995   BIC:                             9156.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept         201.1000      2.446     82.215      0.000     196.300     205.900
race[T.group B]     2.8789      2.969      0.970      0.333      -2.948       8.706
race[T.group C]    -0.8712      2.770     -0.315      0.753      -6.306       4.564
race[T.group D]     0.5360      2.837      0.189      0.850      -5.030       6.102
race[T.group E]     0.6143      3.135      0.196      0.845      -5.538       6.767
==============================================================================
Omnibus:                        0.480   Durbin-Watson:                   1.990
Prob(Omnibus):                  0.787   Jarque-Bera (JB):                0.568
Skew:                          -0.030   Prob(JB):                        0.753
Kurtosis:                       2.900   Cond. No.                         8.57
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [33]:
# Anova table...
from statsmodels.stats.anova import anova_lm
anova_lm(model)

df         sum_sq     mean_sq        F    PR(>F)
race        4.0    1699.671655  424.917914  0.78911  0.532294
Residual  995.0  535785.303345  538.477692      NaN       NaN

Since the Pvalue is greater than Alpha, We Fail to Reject the Ho meaning that the Avg Marks are same across all the categories.

In other words, the Race/Ethinicity has nothing to do with the scores.

#### Perform ANOVA to test the equality of means.

Let us check the normality of the total marks of students from all the groups.

From the above result, we can see that the p-value is greater than 0.05, thus we can say that the total marks of students from each group are normally distributed. Thus the assumption of normality is satisfied.

Let us check the equality of variances.

In [3]:
# perform Levene's test for the equality of variances 


From the above result, we can see that the p-value is greater than 0.05, thus we can say that the population variances are equal for all the samples.

For ⍺ = 0.05 and degrees of freedom (= t-1, N-t) = (4, 995), calculate the critical value.

In [5]:
# calculate the F-value for 95% of confidence level
# use 'stats.f.isf()' to find the F-value corresponding to the upper tail probability 'q'


i.e. if the test statistic value is greater than 2.3809 then we reject the null hypothesis.

In [6]:
# perform one-way ANOVA


We can also use the `anova_lm()` in statsmodels library to perform ANOVA.

In [7]:
# perform one-way ANOVA


#### 2. Ryan is a production manager at an industry manufacturing alloy steels. They have 4 machines - A, B, C and D. Ryan wants to study whether all the machines have equal efficiency. Ryan collects data of tensile strength from all the 4 machines as given. Test at 5% level of significance.

<img src='1_ANOVA.png'>

The null and alternative hypothesis is:

H<sub>0</sub>: The average tensile strength due to all the machines is the same<br>
H<sub>1</sub>: The average tensile strength due to at least one machines is different

In [41]:
# given data
# tensile strength due to machine A
A = [68.7, 75.4, 70.9, 79.1, 78.2]

# tensile strength due to machine B
B = [62.7, 68.5, 63.1, 62.2, 60.3]

# tensile strength due to machine C
C = [55.9, 56.1, 57.3, 59.2, 50.1]

# tensile strength due to machine D
D = [80.7, 70.3, 80.9, 85.4, 82.3]

### One way

In [39]:
# F Critical
stats.f.isf(0.05, dfn = 3, dfd = 16)

3.238871517453585

In [40]:
# F Test
stats.f_oneway(A, B, C, D)

F_onewayResult(statistic=32.03072350199285, pvalue=5.375613532781072e-07)

#### Perform ANOVA to test the equality of means.

Let us check the normality of the tensile strength of all the machines.

In [35]:
# create a dataframe using a dictionary from given data
df_machine = pd.DataFrame({"Machine":["Machine_A", "Machine_B", "Machine_C",
                        "Machine_D"]*5, 
             "Strength":[68.7, 62.7, 55.9, 80.7, 75.4, 68.5,56.1,
                         70.3,70.9,63.1,57.3,80.9,
                         79.1,62.2,59.2,85.4,78.2,60.3,50.1,82.3]})

In [37]:
df_machine.head()

Machine  Strength
0  Machine_A      68.7
1  Machine_B      62.7
2  Machine_C      55.9
3  Machine_D      80.7
4  Machine_A      75.4

In [36]:
# perform Shapiro-Wilk test to test the normality
# shapiro() returns a tuple having the values of test statistics and the corresponding p-value

stats.shapiro(df_machine["Strength"])

ShapiroResult(statistic=0.9503339529037476, pvalue=0.3721875548362732)

From the above result, we can see that the p-value is greater than 0.05, thus we can say that the tensile strengths due to all the machines are normally distributed. Thus the assumption of normality is satisfied.

Let us check the equality of variances.

In [38]:
# perform Levene's test for the equality of variances 
stats.levene(A, B, C, D)

LeveneResult(statistic=0.3969333650936478, pvalue=0.7570021212992085)

From the above result, we can see that the p-value is greater than 0.05, thus we can say that the population variances are equal for all the samples.

Here t (=number of machines) = 4, N (=total observations) = 20 

For ⍺ = 0.05, calculate the Critical Value....

In [42]:
# calculate the F-value for 95% of confidence level
# use 'stats.f.isf()' to find the F-value corresponding to the upper tail probability 'q'
stats.f.isf(0.05, 3, 16)

3.238871517453585

i.e. if the test statistic value is greater than 3.2389 then we reject the null hypothesis.

In [41]:
# perform one-way ANOVA
model = sfa.ols("Strength~Machine", data = df_machine).fit()

anova_lm(model)

df     sum_sq    mean_sq          F        PR(>F)
Machine    3.0  1778.0655  592.68850  32.030724  5.375614e-07
Residual  16.0   296.0600   18.50375        NaN           NaN

The above output shows that the test statistic is greater than 3.2389 and the p-value is less than 0.05. Thus we reject the null hypothesis and conclude that the average tensile strength for **at least one machine is different.**

<a id="post-hoc"></a>
## 2.1 Post-hoc Analysis

If one-way ANOVA rejects the null hypothesis; we conclude that at least one treatment has a different mean. The test does not distinguish a treatment with the different average value. The post-hoc test or `multi comparison test` is used to identify such treatment(s).

In this section, we study the `Tukey's HSD` test. The test calculates the mean difference for each pair of treatments and returns the pair(s) with different average. 

The test statistic of Tukey's HSD test is given as:
<p style='text-indent:28em'> <strong> $T_{\alpha} = q_{{\alpha},(t , f)} \sqrt{\frac{MSE}{n}} $</strong></p>

The value of $q_{{\alpha},(t , f)}$ is obtained from the tukey table.<br>
Where,<br>
t: Number of treatments<br>
f: degrees of freedom for error ($df_{e}$)<br>
MSE: Mean error sum of squares (= $\frac{ESS}{df_{e}}$ = $\sum_{i}^{t}\sum_{j}^{n_{i}}{(x_{ij} - \bar{x_{i}}.)}^{2}$)<br>
n: Number of observations in a treatment

This test is efficient when the sample size for each treatment is equal. If the sample size is not equal fo each treatment then we can use the `Scheffe test`. The `scikit_posthocs.posthoc_scheffe()` can be used to perform the test.

### Example:

#### 1. Ryan is a production manager at an industry manufacturing alloy seals. They have 4 machines - A, B, C and D. Ryan wants to study whether all the machines have equal efficiency. Ryan collects data of tensile strength from all the 4 machines as given. Perform the post-hoc test to find out which machine has a different average. Test at 5% level of significance.

<img src='1_ANOVA.png'>

In [44]:
# use the same dataframe created above
# perform tukey's range test to compare the mean efficiency for pair of machines
# pass the tensile strength to the parameter, 'data'
# pass the name of the machine to the parameter, 'groups'
comp = mc.MultiComparison(data = df_machine['Strength'], 
                          groups = df_machine['Machine'])
# tukey's range test
post_hoc = comp.tukeyhsd()

# print the summary table
post_hoc.summary()

<class 'statsmodels.iolib.table.SimpleTable'>

The `reject=False` for pairs (machine_A, machine_D) and (machine_B, machine_C) denotes that we fail to reject the null hypothesis; and conclude that the average tensile strength due to machine_A and machine_D, machine_B and machine_C is same.

For the pairs (machine_A, machine_B), (machine_A, machine_C), (machine_B, machine_D), and (machine_C, machine_D) the average tensile strength is not the same.

The values in the columns `lower` and `upper` represent the lower and upper bound of the 95% confidence interval for the mean difference. 

### Summary of Tukey HSD Test

A Pairwise Tukey HSD Test is used in Anova to find the difference in the mean which is resulting in variance in the categories.In other words, this test reflects the difference in the performance amongst categories.

<a id="non-para"></a>

## Kruskall Wallis Test

* If one of the assumptions of one-way ANOVA is not satisfied, then we can perform the **Kruskal-Wallis H test which is a non-parametric equivalent test for ANOVA.**

* Wilcoxon Signed-Ranks test is the nonparametric version of the Two Sample Related T-Test

* If the Data is not following Normal Distribution, then one can use Mann Whitney U test inplace of Independent Two Sample T-Test. It is also known as Mann Whitney Wilcoxon Test or the Wilcoxon Rank Sum Test.

A gym trainer wants to provide an energy bar to all his customers to increase muscle strength. Three different companies approached the gym trainer with their high-quality energy bars. The trainer collects an amount of calcium (in g) in the energy bar from three companies and he wants to study whether all the bars have an equal amount of calcium on average. Test at 5% level of significance.

given data:

   * alpha_bar = [24.4, 20.7, 56.9, 19.5]
   * beta_bar = [53.2, 54.7, 20.5, 15.8, 56.6]
   * gamma_bar = [54, 31, 22.8, 24.7]
   
* **Ho: That All the Energy Bars have Same Quantity of Calcium**
* **Ha: That All the Energy Bars Do Not have Same Quantity of Calcium**

In [42]:
# given data
# amount of calcium in alpha_bar
alpha = [24.4, 20.7, 56.9, 19.5]

# amount of calcium in beta_bar
beta = [53.2, 54.7, 20.5, 15.8, 56.6]

# amount of calcium in gamma_bar
gamma = [54, 31, 22.8, 24.7]

In [43]:
# Check Normality
stats.shapiro(alpha)

ShapiroResult(statistic=0.7282745838165283, pvalue=0.023655038326978683)

In [44]:
# Check Variance....
stats.levene(alpha,beta,gamma)

LeveneResult(statistic=0.16749212708722086, pvalue=0.8481074994770258)

In [45]:
stats.kruskal(alpha,beta,gamma)   # for more than 2

KruskalResult(statistic=0.22747252747252844, pvalue=0.8924933076960211)

In [46]:
# Fail to reject

In [47]:
stats.mannwhitneyu(alpha,gamma)   # for 2 

MannwhitneyuResult(statistic=5.0, pvalue=0.4857142857142857)